Source/Core/Common/BitFieldView.h

```diff
@@ -0,0 +1,788 @@
+
+#include "Common/Assert.h"    // DEBUG_ASSERT
+#include "Common/BitUtils.h"  // Common::BitCast
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Assertation helper classes
+
+template <typename host_t>
+struct AssertHostTypeIsSane
+{
+  AssertHostTypeIsSane() = default;
+  static_assert(std::is_integral_v<host_t> || std::is_enum_v<host_t>,
+                "Given host type is not sane.  You must define a partial specialization for it.");
+};
+template <typename field_t>
+struct AssertFieldTypeIsSane
+{
+  AssertFieldTypeIsSane() = default;
+  static_assert(std::is_integral_v<field_t> || std::is_enum_v<field_t>,
+                "Given field type is not sane.  It must be an integral or enumerated type.");
+};
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+struct AssertBitFieldFixedIsSane
+{
+  AssertBitFieldFixedIsSane() = default;
+  static_assert(start <= 8 * sizeof(host_t), "BitField out of range");
+  static_assert(start + width <= 8 * sizeof(host_t), "BitField out of range");
+  static_assert(width <= 8 * sizeof(field_t), "Field is too wide");
+  static_assert(width > 0, "BitField is invalid (zero width)");
+  static_assert(!std::is_same_v<field_t, bool> || width == 1,
+                "Boolean fields should only be 1 bit wide");
+};
+template <typename field_t, std::size_t start, std::size_t width, std::size_t length,
+          typename host_t>
+struct AssertBitFieldFixedArrayIsSane
+{
+  AssertBitFieldFixedArrayIsSane() = default;
+  static_assert(start <= 8 * sizeof(host_t), "BitFieldArray out of range");
+  static_assert(start + width * length <= 8 * sizeof(host_t), "BitFieldArray out of range");
+  static_assert(width <= 8 * sizeof(field_t), "Field is too wide");
+  static_assert(width > 0, "Field is invalid (zero width)");
+  static_assert(!std::is_same_v<field_t, bool> || width == 1,
+                "Boolean fields should only be 1 bit wide");
+};
+template <typename field_t, typename host_t>
+struct AssertBitFieldIsSane
+{
+  AssertBitFieldIsSane() = delete;
+  AssertBitFieldIsSane(const std::size_t start, const std::size_t width)
+  {
+    DEBUG_ASSERT(start <= 8 * sizeof(host_t));          // BitField out of range
+    DEBUG_ASSERT(start + width <= 8 * sizeof(host_t));  // BitField out of range
+    DEBUG_ASSERT(width <= 8 * sizeof(field_t));         // BitField is too wide
+    DEBUG_ASSERT(width > 0);                            // BitField is invalid (zero width);
+    if constexpr (std::is_same_v<field_t, bool>)
+      DEBUG_ASSERT(width == 1);  // Boolean fields should only be 1 bit wide
+  }
+};
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Simple, functional programming.
+
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+constexpr field_t GetBitFieldFixed(const host_t host)
```

Comment:
Thinking about this harder, is there any reason to use this over, say `Common::ExtractBits`?

Reply:
It looks like Common::ExtractBits always produces an unsigned result, so it is not a perfect shoe-in.  The irony of BitFieldView.h and BitField.h is that they both hinge on implementation-defined behavior (arithmetic right-shift w/ sign-extension) to eliminate implementation-defined behavior (C bit structs).  That being said, I do agree that it appears BitField.h (2014) and BitUtils.h (2017) invented nearly identical bit-extraction functions independent of one another, and it would be better to not repeat ourselves now that we've recognized it.  The only issue is finding a arbitrary bit-width sign-extension technique which is performant.  Do you have any suggestions?

Reply:
<https://godbolt.org/z/KoW8jfKP4>

Reply:
So it seems the heart of BitFieldView, it's Getter/Setter functions, really belong in BitUtils.h, with BitFieldView.h shall simply contain interface classes for for pretty syntax.  Oh, how the mighty has fallen.